egui/src/widgets/text_edit/filter.rs

```diff
@@ -0,0 +1,178 @@
+use super::{Action, InputData};
+
+use std::{borrow::Cow, cmp};
+
+/// A type that filters user text input.
+pub trait FilterInput {
+    /// Determine how to treat the user `input` based on the current
+    /// value of `InputData`.
+    fn filter_input(data: InputData<'_>) -> Action;
+}
+
+impl FilterInput for String {
+    fn filter_input(data: InputData<'_>) -> Action {
+        Action::Insert(data.input)
+    }
+}
+
+impl FilterInput for str {
+    fn filter_input(data: InputData<'_>) -> Action {
+        Action::Insert(data.input)
+    }
+}
+
+macro_rules! impl_filter_input_unsigned {
+    ($type: ty, $max_str: expr) => {
+        impl FilterInput for $type {
+            fn filter_input(data: InputData<'_>) -> Action {
+                let InputData { buffer, input, .. } = data;
+                let max_str = $max_str;
+                let max_str_len = max_str.len();
+
+                // TODO should we assume that the user did not add data
+                // to the buffer from outside the UI?
+
+                // Filter out all the non-numeric characters.
+                let filter = input.as_str().replace(|c: char| !c.is_numeric(), "");
+
+                let new_size = buffer.len() + filter.len();
+
+                match new_size.cmp(&max_str_len) {
+                    // Guaranteed to be below the limit so we insert.
+                    cmp::Ordering::Less => Action::Insert(filter),
+                    // Guaranteed to exceed the limit so we overwrite.
+                    cmp::Ordering::Greater => Action::Overwrite(Cow::Borrowed(max_str)),
+                    // Slow case where we need to actually compare with the limit.
+                    cmp::Ordering::Equal => {
+                        // Note that we don't parse the strings because its too slow.
```

Comment:
This is not a great motivation IMHO. We only need to run this code when new text is entered, so it is not even every frame.